### PR TITLE
fix deadlock after reconnection

### PR DESCRIPTION
**Pull request by marcin-zyrkowski**
Those changes fix the deadlock state that Client could go while working in the poor radio (LTE) conditions.

Changelog: NONE

### CODE REVIEW

File: src/mender/client/http_requests.py [hunk 1]
```diff
@@ -28,6 +28,7 @@ def http_request(method, server_url, **kwargs):
         requests.URLRequired,
         requests.TooManyRedirects,
         requests.Timeout,
+        requests.ReadTimeout,
```
Comment (oleorhagen) [inline src/mender/client/http_requests.py:31]: Question:
Why is this needed?
pydoc gives me:
```
Help on class Timeout in requests:

requests.Timeout = class Timeout(RequestException)
 |  requests.Timeout(*args, **kwargs)
 |
 |  The request timed out.
 |
 |  Catching this error will catch both
 |  :exc:`~requests.exceptions.ConnectTimeout` and
 |  :exc:`~requests.exceptions.ReadTimeout` errors.
```
So, the `Timeout` should already catch this, no?

File: src/mender/statemachine/statemachine.py [hunk 2]
```diff
@@ -364,8 +364,14 @@ def run(self, context):
 class ArtifactFailure(State):
     def run(self, context):
         log.info("Running the ArtifactFailure state...")
-        # return _UpdateDone()
-        raise UnsupportedState("ArtifactFailure is unhandled by the API client")
+        log.warning(
+            "The Artifact has not been properly downloaded due to lack of Internet access or a server failure."
+        )
+        log.warning(
+            "The next attemp of the update will happen after the UpdatePollIntervalSeconds config variable."
```
[the remark this answers is not in the document]
Comment (marcin-zyrkowski) [reply src/mender/statemachine/statemachine.py:371]: I will fix, thanks.

File: src/mender/statemachine/statemachine.py [hunk 3]
```diff
@@ -364,8 +364,14 @@ def run(self, context):
 class ArtifactFailure(State):
     def run(self, context):
         log.info("Running the ArtifactFailure state...")
-        # return _UpdateDone()
-        raise UnsupportedState("ArtifactFailure is unhandled by the API client")
+        log.warning(
+            "The Artifact has not been properly downloaded due to lack of Internet access or a server failure."
```
[the remark this answers is not in the document]
Comment (marcin-zyrkowski) [reply src/mender/statemachine/statemachine.py:368]: Ageed. Will move to the Download state.

File: src/mender/statemachine/statemachine.py [hunk 4]
```diff
@@ -387,6 +393,13 @@ def __init__(self):
         self.current_state = Download()
 
     def run(self, context):
+        """Here goes a hack: it is assumed that download would succeded
+         and the state would change to ArtifactInstall and that would do other things
+         in external system scripts while this client exits.
+         Still in real world scenario the download is failing
+         and changes the state to ArtifactFailure.
+         After returning from ArtifactFailure state we will need to return to IdleState
+         in AuthorizedStateMachine to let it go another chance after some time."""
```
[the remark this answers is not in the document]
Comment (marcin-zyrkowski) [reply src/mender/statemachine/statemachine.py:402]: Agreed. Will remove the word 'hack'.

File: src/mender/client/deployments.py [hunk 5]
```diff
@@ -212,6 +220,10 @@ def download_and_resume(
                 headers=req_headers,
                 stream=True,
                 verify=server_certificate or True,
+                timeout=(
```
[the remark this answers is not in the document]
Comment (marcin-zyrkowski) [reply src/mender/client/deployments.py:223]: Well, I agree that add the default Timeout to the other http_request is a good idea. Still there were no issue reported regarding the requests. If added then there should be some work required on the 'good' default values, the experiments, the testing... I would avoid this in the stage of this specific issue.

File: src/mender/client/deployments.py [hunk 6]
```diff
@@ -223,25 +235,70 @@ def download_and_resume(
                 log.debug(f"Opening file to write at offset {offset}")
                 with open(artifact_path, "rb+") as fh:
                     fh.seek(offset)
+                    date_past = datetime.now()
+                    date_start = datetime.now()
                     for data in response.iter_content(
-                        chunk_size=1024 * 1024
+                        chunk_size=DOWNLOAD_CHUNK_SIZE_BYTES
                     ):  # 1MiB at a time
```
[the remark this answers is not in the document]
Comment (marcin-zyrkowski) [reply src/mender/client/deployments.py:242]: I changes the comment to: "# 1 chunk at a time".
I propose to keep the smaller chunk to increase probability that the download would be successful. Having a copper or fiber connection the client would get a greet speed and robustness. However using LTE one may have much smaller troughputs like 50Kbit up to 2Mbits. Having in mind that we have the mechanism for retransmissions we just adjusting to the link conditions.

[the remark this answers is not in the document]
Comment (marcin-zyrkowski) [reply src/mender/client/deployments.py:242]: Generally it would be good to extract maybe that chunk size to the config file.

File: src/mender/client/http_requests.py (hunk 1 above)
Comment (marcin-zyrkowski) [reply src/mender/client/http_requests.py:31]: That's a mistery for me. I had to add that exception because it was thrown and not catched during my testing. Similarily I had to add the separate section in statemachine.py lines 280-288 to catch ConnectionError despite that MenderRequestsException section should catch this ConnectionError in the section just above ...

File: src/mender/client/deployments.py [hunk 7]
```diff
@@ -223,25 +235,70 @@ def download_and_resume(
...
-            log.debug(f"Got Error. Wrote {offset} bytes. Total is {content_length}.")
+            log.debug(e)
+            t_difference = datetime.now() - date_start
+            t_diff_millis = (
+                t_difference.seconds * 1000 + t_difference.microseconds / 1000
+            )
+            log.debug(
+                f"Got Error. Wrote {offset} bytes. Total is {content_length}. Time {t_diff_millis:.0f} milliseconds"
+            )
+        except requests.ConnectionError as e:
```
[the remark this answers is not in the document]
Comment (marcin-zyrkowski) [reply src/mender/client/deployments.py:280]: For some reason it didn't catch in the wrapper. It was unhandled during my testing. I just had to add the catch for ConnectionError despite that MenderRequestsException section should catch this ConnectionError in the section just above ...

File: src/mender/client/deployments.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply src/mender/client/deployments.py:242]: Agreed.

File: src/mender/client/deployments.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply src/mender/client/deployments.py:280]: I see. Hmm odd :thinking:

File: src/mender/client/http_requests.py (hunk 1 above)
Comment (oleorhagen) [reply src/mender/client/http_requests.py:31]: Yeh, I've seen. It's odd

**Comment by oleorhagen** [comment]
Superseeded by #112
